test/integration/specs/kbaseCellToolbarMenu_test.js

```diff
@@ -0,0 +1,327 @@
+/*global describe, it, browser, afterEach, beforeEach*/
+/* eslint {strict: ['error', 'global']} */
+'use strict';
+
+const {login, openNarrative, clickWhenReady} = require('../wdioUtils.js');
+
+/*
+Notes:
+
+The test narrative has 5 markdown cells. The first cell has summary text, the remaining 4 contain the text "Cell #".
+The first three cells are collapsed, the last two are expanded.
+
+Tests currently cover common cases of cell movement; additional cases can be added, e.g. try to move 
+past the end, or before the beginning, try moving the same cell multiple times, inspect the state
+of all cells after a movement.
+
+This test suite could be expanded to cover all usage of the cell toolbar (kbaseCellToolbarMenu.js), or they
+could be in separate files.
+
+*/
+
+const testData = {
+    common: {
+        unselectedBorder: '1px 1px 1px 5px solid rgb(204, 204, 204)',
+        selectedBorder: '1px 1px 1px 5px solid rgb(75, 184, 86)',
+    },
+    all: {
+        TEST_CASE1: {
+            cellIndex: 1,
+            title: 'Narrative Cell Toolbar Testing'
+        },
+        TEST_CASE2: {
+            cellIndex: 2,
+            title: 'Cell 1'
+        },
+        TEST_CASE3: {
+            cellIndex: 3,
+            title: 'Cell 2'
+        },
+        TEST_CASE4: {
+            cellIndex: 3,
+            title: 'Cell 2'
+        },
+        TEST_CASE5: {
+            cellIndex: 4,
+            body: 'Cell 3'
+        }
+    },
+    ci: {
+        TEST_CASE1: {
+            narrativeId: 58675,
+        },
+        TEST_CASE2: {
+            narrativeId: 58675,
+        },
+        TEST_CASE3: {
+            narrativeId: 58675,
+        },
+        TEST_CASE4: {
+            narrativeId: 58675,
+        },
+        TEST_CASE5: {
+            narrativeId: 58675,
+        }
+    },
+    'narrative-dev': {
+        TEST_CASE1: {
+            narrativeId: 80970,
+        },
+        TEST_CASE2: {
+            narrativeId: 80970,
+        },
+        TEST_CASE3: {
+            narrativeId: 80970,
+        },
+        TEST_CASE4: {
+            narrativeId: 80970,
+        },
+        TEST_CASE5: {
+            narrativeId: 80970,
+        }
+    }
+};
+
+const testCases = mergeObjects([testData[browser.config.testParams.ENV], testData.all]);
```

Comment:
Wouldn't it be simpler to do something like:

``` js
const narrativeIdsByEnv = {
  ci: 58675,
  'narrative-dev': 80970,
},
narrativeEnv = browser.config.testParams.ENV;
if (!narrativeIdsByEnv[narrativeEnv]) {
// bail early 
}
const narrativeId = narrativeIdsByEnv[narrativeEnv];

// in the tests
it('moves a minimized selected cell down', async () => {
  const testCase = testCases.TEST_CASE1;
  await login();
  const narrativeContainer = await openNarrative(narrativeId);
  // etc.
```

?

Reply:
An update to the PR will include a general solution to this.
I had punted on this, because I wanted to keep the test data confined to a single json-compatible structure -- probably better to keep the test data in a file at some point, as it can get pretty large for some test cases, and keeping it in the set file, while simpler for now, makes the test files bigger.

Reply:
Thinking about this a bit more and looking at the other existing integration tests (which aren't in the `truss` / bulk import cell branch yet), I think it would make sense to have a utility function to get the narrative ID for a given test in a certain environment, and to skip the tests if there isn't a narrative available for that test in that env. Something like

```js
// in a utils file
function getTestNarrativeId(testId) {
  const tests = {
    kbaseCellToolbarMenu: {
      ci: 58675,
      'narrative-dev': 80970,
    },
    kbaseCellSidebarMenu: {
      ci: 12345,
    },
    ...
  };

  const env = browser.config.testParams.ENV;
  if (!tests[testId]) {
    throw new Error(`Invalid testId ${testId}`);
  }
  if (!tests[testId][env]) {
    return null;
  }
  return tests[testId][env];
}
```

```js
// in the test file
const testId = 'kbaseCellToolbarMenu'; // could derive dynamically from the file name
const narrativeId = testUtils.getTestNarrativeId(testId);
if (!narrativeId) {
  // skip the tests - no narrative available in this env
}
```

That would save from having repeated boilerplate in each integration test. 

(Going a step further, you could enclose the tests in a function, and have a `runTests` util function to which you pass the test ID and the test function; `runTests` could either execute the function (thereby running the tests) or could skip that test set if there was no narrative available.)

Reply:
The idea was to eventually move all test data into a separate json file, so they all live in a single test data object.
The test data has since been moved into a companion test data file.

Reply:
I'm sure the integration test support will continually evolve. There are few enough tests that it is easy to refactor them all in a few minutes when we do change that support.

Since the original PR, I refactored the narrative-centric util methods into a NarrativeTesting class, which knows about the structure of test data utilized by the tests I've written. An instance is created for each test, given the test data, a test case key, the environment. Some things are assumed to be in the global test config, which I'm not comfortable with ... but it is a process.

(BTW some of my comments may have come after yours but not reflect them, because this page was stale.)

Actually, globally, I'm not sure that this is the best testing framework to use anyway. It is actually quite easy to create false positive test just by missing an await ... the promises are uncontrolled, and the test completes without error, although the promises may actual resolve to an error. It is asking a lot for devs to remember when they need to use await (which is just about always when using the WebdriverIO api).

I found earlier that the synchronous tests did not work well, especially if one needs to wrap them and mix in custom test code, but that may have changed since, and I may not have tried hard enough.